ansible_base/authentication/utils/claims.py

```diff
@@ -330,39 +347,61 @@ def process_organization_and_team_memberships(results) -> None:
                 new_team, _ = Team.objects.get_or_create(name=team_name, organization_id=org_id)
 
 
+# NOTE(cutwater): Current class is sub-optimal, since it loads the data that has been already loaded
+#  at the teams and organizations creation. Next step will be combining teams and organizations creation with
+#  this class and transforming it into a reconciliation use case class. This implies either
+#  removal or update of a pluggable interface.
 class ReconcileUser:
 
-    @staticmethod
-    def reconcile_user_claims(user: Optional[AbstractUser], authenticator_user: AuthenticatorUser) -> None:
+    @classmethod
+    def _remove_role_from_user(cls, user: AbstractUser, rd: RoleDefinition):
+        for role in user.has_roles.filter(role_definition=rd):
+            rd.remove_permission(user, role.content_object)
 
+    @classmethod
+    def reconcile_user_claims(cls, user: AbstractUser, authenticator_user: AuthenticatorUser) -> None:
         logger.info("Reconciling user claims")
-        claims = getattr(user, 'claims', getattr(authenticator_user, 'claims'))
+
+        claims = getattr(user, 'claims', authenticator_user.claims)
+        org_member_rd = RoleDefinition.objects.get(name=constants.ORGANIZATION_MEMBER_ROLE_NAME)
+        team_member_rd = RoleDefinition.objects.get(name=constants.TEAM_MEMBER_ROLE_NAME)
+
+        remove_users = authenticator_user.provider.remove_users
+        if remove_users:
+            cls._remove_role_from_user(user, org_member_rd)
+            cls._remove_role_from_user(user, team_member_rd)
+
+        org_names = claims['organization_membership'].keys() | claims['team_membership'].keys()
+        orgs_by_name = {org.name: org for org in Organization.objects.filter(name__in=org_names)}
 
         for org_name, is_member in claims['organization_membership'].items():
+            org = orgs_by_name.get(org_name)
+            if org is None:
+                logger.info("Skipping organization '%s'", org_name)
```

Comment:
I might change this to add "because the org does not exist" or something to indicate why we skipped.